Update fuse docs for Podman 5.x (#2842)
codenvy/che-docs: modules/administration-guide/pages/enabling-fuse-for-all-workspaces.adoc

[id="enabling-fuse-overlayfs-for-all-workspaces"]
= Enabling fuse-overlayfs for all workspaces

> For Podman 5.x, the `/home/user/.config` must be owned by the current user for Podman to function correctly.
> The `storage.conf` file for Podman is typically stored in this folder.
> This document explains how to configure the workspace's container entrypoint script for the workspace so that fuse-overlayfs is being used for all workspaces using that container.
> The Universal Developer Image (UDI) already contains the necessary configuration by default.

Reviewer: ```suggestion
Learn about configuring the workspace's container entrypoint script so that fuse-overlayfs is used for all workspaces using that container.

The Universal Developer Image (UDI) already contains the necessary configuration by default. However, you must configure the script manually if you use custom images.
```

.Prerequisites

> * For OpenShift versions 4.14 and lower, the xref:administration-guide:enabling-access-to-dev-fuse-for-openshift.adoc[] section has been completed.
> 
> * An active `{orch-cli}` session with administrative permissions to the destination OpenShift cluster. See {orch-cli-link}.

Reviewer: ```suggestion
* An active `{orch-cli}` session with administrative permissions to the destination OpenShift cluster. See {orch-cli-link}.

* For OpenShift versions 4.14 and lower, the xref:administration-guide:enabling-access-to-dev-fuse-for-openshift.adoc[] section has been completed.

* For Podman 5.x, the `/home/user/.config` must be owned by the current user. This is where the `storage.conf` file for Podman is typically stored.
```

Author: > For Podman 5.x, the `/home/user/.config` must be owned by the current user. This is where the `storage.conf` file for Podman is typically stored.

This part is what the procedure aims to do, so it doesn't belong in the prerequisites. I've added a new commit to explain this part in the procedure steps: 5795e7c957cf887fa2d12048ed9fcace6ace36f8

WDYT?

Reviewer: I see. sorry about that.

Reviewer: your edit looks great. now I understand how it all works.

[id="enabling-fuse-overlayfs-for-all-workspaces"]
= Enabling fuse-overlayfs for all workspaces

For Podman 5.x, the `/home/user/.config` must be owned by the current user for Podman to function correctly.
The `storage.conf` file for Podman is typically stored in this folder.

> This document explains how to configure the container entrypoint script for the workspace so that fuse-overlayfs is being used.

Reviewer: after configuration, is fuse-overlayfs used by the workspace with the configured script? the wording is a bit unclear, especially with "so that fuse-overlayfs is being used."

Author: The script is configured for the container image entrypoint. After that, a new container image is built, and whatever workspace uses that container image, will be using fuse-overlay fs for podman

> The Universal Developer Image (UDI) already contains the necessary configuration by default.

Reviewer: when you refer to the configuration already contained in UDI, what sort of configuration are you referring to? the one from the previous line (container entrypoint script)?

Author: Yes, the container entrypoint script, for example:
https://github.com/devfile/developer-images/blob/e24cbd7d58b8a744467db2e404314b63c30d8f85/base/ubi9/entrypoint.sh#L8-L16

.Prerequisites

> * The xref:administration-guide:enabling-access-to-dev-fuse-for-openshift.adoc[] section has been completed. This is not required for OpenShift versions 4.15 and later.

Reviewer: i'm wondering if it would make sense to reword the prerequisite since it's only necessary for older OCP versions.

maybe we should add FOR OCP VERSIONS 4.14 AND LOWER there?

Author: I've updated the wording:
```
For OpenShift versions 4.14 and lower, the xref:administration-guide:enabling-access-to-dev-fuse-for-openshift.adoc[]
section has been completed.
```

+
[NOTE]
====
For OpenShift versions 4.14 and lower, the `io.openshift.podman-fuse: ""` annotation is also required.
====

> . [OPTIONAL] If you are using a custom image for the workspace container, create the `/home/user/.config` folder and configure the `storage.conf` file on runtime via the entrypoint.

Reviewer: ```suggestion
. Optional: If you are using a custom image for the workspace container, create the `/home/user/.config` folder and configure the `storage.conf` file on runtime via the entrypoint.
```

+
[NOTE]
====
For OpenShift versions 4.14 and lower, the `io.openshift.podman-fuse: ""` annotation is also required.
====

> . Optional: If you are using a custom image for the workspace container, create the `/home/user/.config` folder and configure the `storage.conf` file on runtime via the entrypoint.

Reviewer: ```suggestion
. [OPTIONAL] If you are using a custom image for the workspace container, create the `/home/user/.config` folder and configure the `storage.conf` file on runtime via the entrypoint.
```

Author: I've applied this change, but the doc renders this as:
![image](https://github.com/user-attachments/assets/2c094e2e-94d8-4201-98e7-233e029ddb1e)

which I don't think is expected

Reviewer: oh no, that's definitely not what I intended. I'll have to check what happened.

Reviewer: I mashed together optional step and admonition formatting and created this abomination.

To do this, add the following to the workspace container image's entrypoint script before building the image.
+
====
[source,subs="+quotes,+macros"]
----
# Configure container builds to use vfs or fuse-overlayfs
if [ ! -d "${HOME}/.config/containers" ]; then
  mkdir -p ${HOME}/.config/containers
  if [ -c "/dev/fuse" ] && [ -f "/usr/bin/fuse-overlayfs" ]; then
    (echo '[storage]';echo 'driver = "overlay"';echo '[storage.options.overlay]';echo 'mount_program = "/usr/bin/fuse-overlayfs"') > ${HOME}/.config/containers/storage.conf
  else
    (echo '[storage]';echo 'driver = "vfs"') > "${HOME}"/.config/containers/storage.conf
  fi
fi
----
====
+
This ensures that if the `/home/user/.config` doesn't already exist, the folder is created and owned by `user`.
The `/home/user/.config` may already exist for example, if it was stored in a persistent volume.
+

> [NOTE]
> ====
> This is configured in the UDI by default. Therefore this step is only required if you are using a custom image for the workspace container.
> ====

Reviewer: ```suggestion

```

For OpenShift versions before 4.15, the `io.openshift.podman-fuse: ""` annotation is also required.
====

. If you are using a custom image for the workspace container, create the `/home/user/.config` folder and configure the `storage.conf` file on runtime via the entrypoint.

> To do this, create add the following to the workspace container image's entrypoint script before building the image.

Reviewer: create and add?

Author: Updated, it should only be `add`

+
====

> [source,bash,subs="+quotes,+macros"]

Reviewer: please don't use bash as a source language, it causes issues https://redhat-documentation.github.io/supplementary-style-guide/#code-example-syntax-highlighting

Author: Fixed

----
# Configure container builds to use vfs or fuse-overlayfs
if [ ! -d "${HOME}/.config/containers" ]; then
  mkdir -p ${HOME}/.config/containers
  if [ -c "/dev/fuse" ] && [ -f "/usr/bin/fuse-overlayfs" ]; then
    (echo '[storage]';echo 'driver = "overlay"';echo '[storage.options.overlay]';echo 'mount_program = "/usr/bin/fuse-overlayfs"') > ${HOME}/.config/containers/storage.conf
  else
    (echo '[storage]';echo 'driver = "vfs"') > "${HOME}"/.config/containers/storage.conf
  fi
fi
----
====
+
This ensures that if the `/home/user/.config` doesn't already exist, the folder is created and owned by `user`.
The `/home/user/.config` may already exist for example, if it was stored in a persistent volume.
+
[NOTE]
====

> This is configured in the UDI by default. Therefore this step is only required if you are using a custom image for the workspace container.

Reviewer: maybe we could mark the whole step as [OPTIONAL] since it's only required for custom images

Author: I've added an `Optional:` at the start of the sentence, does that work?
